cmd/swarm/swarm-snapshot/create.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of go-ethereum.
+//
+// go-ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// go-ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with go-ethereum. If not, see <http://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/cmd/utils"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p/simulations"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/swarm/network"
+	"github.com/ethereum/go-ethereum/swarm/network/simulation"
+	cli "gopkg.in/urfave/cli.v1"
+)
+
+var serviceFuncs = adapters.Services{
+	"discovery": newService,
+}
+
+const testMinProxBinSize = 2
+const NoConnectionTimeout = 1
+
+var discoveryEnabled = true
+
+func init() {
+	adapters.RegisterServices(serviceFuncs)
+}
+
+func create(ctx *cli.Context) error {
+	log.PrintOrigins(true)
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(verbosity), log.StreamHandler(os.Stdout, log.TerminalFormat(true))))
+	if len(ctx.Args()) < 1 {
+		return errors.New("argument should be the filename to verify or write-to")
+	}
+	filename = ctx.Args()[0]
+	err := ResolvePath()
+	if err != nil {
+		return err
+	}
+	err = discoverySnapshot(10, adapters.NewSimAdapter(serviceFuncs))
+	if err != nil {
+		utils.Fatalf("Simulation failed: %s", err)
+	}
+
+	return err
+}
+
+func discoverySnapshot(nodes int, adapter adapters.NodeAdapter) error {
+	// create network
+	net := simulations.NewNetwork(adapter, &simulations.NetworkConfig{
+		ID:             "0",
+		DefaultService: "discovery",
+	})
+	defer net.Shutdown()
+	ids, err := net.AddNodes(nodes)
+
+	if err != nil {
+		return err
+	}
+
+	events := make(chan *simulations.Event)
+	sub := net.Events().Subscribe(events)
+	select {
+	case ev := <-events:
```

Comment:
We should test that we get exactly the `node` events we are expecting.

Reply:
by that you mean that we should listen on connect events when initiating the topology connect methods and looking through the peer pot map for the the connections that should be made?

Reply:
this is not a test

Reply:
@zelig WDYM? this checks that no connection events are fired between the nodes

Reply:
@nolash, somewhere in the back of my mind lays the vague memory of you asking explicitly to check that _no connections are formed between nodes after starting them_. this is why I added this code block.
Regarding the check for `just the connections we want` - wouldn't the `ConnectNodesChain` below cause possibly undesired connections? i.e. the initial connections would not necessarily match what is to be matched in the PPMap, in turn causing tests to flake?
I'd need to hack around this to actually make the hive connect loop to actually wait on a signal to start discovery. Not sure how clean this would be.
Please advise

Reply:
Right, I remember.

I think the working assumption right now is that we don't care if the connections start earlier.